### PR TITLE
cudf resolve nvcc 11.0 compiler crashes during codegen

### PR DESCRIPTION
**Pull request by robertmaynard**
When compiling with nvcc 11.0 any translation unit that uses arrow::AllocateBuffer or arrow::AllocateBitmap would cause a
compiler crash during codegen.

To avoid this issue we introduce a C++ shim between cudf/interop and arrow that does the arrow allocations for us.

### CODE REVIEW

File: cpp/src/interop/detail/arrow_allocator.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
Comment (ttnghia) [inline cpp/src/interop/detail/arrow_allocator.cpp:2]:
```suggestion
 * Copyright (c) 2021, NVIDIA CORPORATION.
```

File: cpp/src/interop/detail/arrow_allocator.hpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
Comment (ttnghia) [inline cpp/src/interop/detail/arrow_allocator.hpp:2]:
```suggestion
 * Copyright (c) 2021, NVIDIA CORPORATION.
```

**Review by ttnghia — APPROVED** [review]
Thanks for working on this :)

**Comment by harrism** [comment]
I assume the crash is fixed in more recent nvcc? If not, we should file an nvcc bug.

File: cpp/src/interop/detail/arrow_allocator.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include <cudf/detail/interop.hpp>
+
+namespace cudf {
+namespace detail {
+
+std::unique_ptr<arrow::Buffer> allocate_arrow_buffer(const int64_t size, arrow::MemoryPool* ar_mr)
+{
+  /*
+  nvcc 11.0 generates Internal Compiler Error during codegen when arrow::AllocateBuffer
+  and `ValueOrDie` or are used inside a CUDA compilation unit.
```
Comment (harrism) [inline cpp/src/interop/detail/arrow_allocator.cpp:26]:
```suggestion
  and `ValueOrDie` are used inside a CUDA compilation unit.
```

File: cpp/src/interop/detail/arrow_allocator.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  auto result = arrow::AllocateBuffer(size, ar_mr);
+  CUDF_EXPECTS(result.ok(), "Failed to allocate Arrow buffer");
+  return std::move(result).ValueOrDie();
+}
+
+std::shared_ptr<arrow::Buffer> allocate_arrow_bitmap(const int64_t size, arrow::MemoryPool* ar_mr)
+{
+  /*
+  nvcc 11.0 generates Internal Compiler Error during codegen when arrow::AllocateBuffer
+  and `ValueOrDie` or are used inside a CUDA compilation unit.
```
Comment (harrism) [inline cpp/src/interop/detail/arrow_allocator.cpp:40]:
```suggestion
  and `ValueOrDie` are used inside a CUDA compilation unit.
```

**Review by harrism — APPROVED** [review]

**Comment by robertmaynard** [comment]
> I assume the crash is fixed in more recent nvcc? If not, we should file an nvcc bug.

You are correct. nvcc 11.1 fixed the crash

**Review by codereport — APPROVED** [review]

**Comment by codereport** [comment]
@gpucibot merge

**Pull request merged**